<a href="https://colab.research.google.com/github/Manal-L/implementez_modele_scoring/blob/main/projet7_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


Implémentez un modèle de scoring

#Chargement des données

In [1]:
import pandas as pd
import zipfile
from google.colab import drive

drive.mount('/content/drive')

#le chemin vers le fichier
file_path = '/content/drive/My Drive/df_clean.zip'

with zipfile.ZipFile(file_path, 'r') as zip_ref:
    zip_ref.extractall('/content/')

Mounted at /content/drive


In [2]:
import os
os.chdir('/content/drive/My Drive/MLflow')


In [3]:
!ls

mlruns	projet7_1.ipynb


In [6]:
df = pd.read_csv('/content/df_clean.csv')
print(df.shape)
df.head()

(356251, 770)


SK_ID_CURR  TARGET  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  \
0      100002     1.0            0             0                0   
1      100003     0.0            1             0                1   
2      100004     0.0            0             1                0   
3      100006     0.0            1             0                0   
4      100007     0.0            0             0                0   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE  \
0             0          202500.0    406597.5      24700.5         351000.0   
1             0          270000.0   1293502.5      35698.5        1129500.0   
2             0           67500.0    135000.0       6750.0         135000.0   
3             0          135000.0    312682.5      29686.5         297000.0   
4             0          121500.0    513000.0      21865.5         513000.0   

   ...  CC_NAME_CONTRACT_STATUS_Sent proposal_MAX  \
0  ...                                        NaN   
1  ...                                        NaN   
2  ...                                        NaN   
3  ...                                      False   
4  ...                                        NaN   

   CC_NAME_CONTRACT_STATUS_Sent proposal_MEAN  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         0.0   
4                                         NaN   

   CC_NAME_CONTRACT_STATUS_Sent proposal_SUM  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        0.0   
4                                        NaN   

   CC_NAME_CONTRACT_STATUS_Sent proposal_VAR  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        0.0   
4                                        NaN   

   CC_NAME_CONTRACT_STATUS_Signed_MIN  CC_NAME_CONTRACT_STATUS_Signed_MAX  \
0                                 NaN                                 NaN   
1                                 NaN                                 NaN   
2                                 NaN                                 NaN   
3                               False                               False   
4                                 NaN                                 NaN   

   CC_NAME_CONTRACT_STATUS_Signed_MEAN  CC_NAME_CONTRACT_STATUS_Signed_SUM  \
0                                  NaN                                 NaN   
1                                  NaN                                 NaN   
2                                  NaN                                 NaN   
3                                  0.0                                 0.0   
4                                  NaN                                 NaN   

   CC_NAME_CONTRACT_STATUS_Signed_VAR  CC_COUNT  
0                                 NaN       NaN  
1                                 NaN       NaN  
2                                 NaN       NaN  
3                                 0.0       6.0  
4                                 NaN       NaN  

[5 rows x 770 columns]

In [ ]:
list(df)

['SK_ID_CURR',
 'TARGET',
 'CODE_GENDER',
 'FLAG_OWN_CAR',
 'FLAG_OWN_REALTY',
 'CNT_CHILDREN',
 'AMT_INCOME_TOTAL',
 'AMT_CREDIT',
 'AMT_ANNUITY',
 'AMT_GOODS_PRICE',
 'REGION_POPULATION_RELATIVE',
 'DAYS_BIRTH',
 'DAYS_EMPLOYED',
 'DAYS_REGISTRATION',
 'DAYS_ID_PUBLISH',
 'OWN_CAR_AGE',
 'FLAG_MOBIL',
 'FLAG_EMP_PHONE',
 'FLAG_WORK_PHONE',
 'FLAG_CONT_MOBILE',
 'FLAG_PHONE',
 'FLAG_EMAIL',
 'CNT_FAM_MEMBERS',
 'REGION_RATING_CLIENT',
 'REGION_RATING_CLIENT_W_CITY',
 'HOUR_APPR_PROCESS_START',
 'REG_REGION_NOT_LIVE_REGION',
 'REG_REGION_NOT_WORK_REGION',
 'LIVE_REGION_NOT_WORK_REGION',
 'REG_CITY_NOT_LIVE_CITY',
 'REG_CITY_NOT_WORK_CITY',
 'LIVE_CITY_NOT_WORK_CITY',
 'EXT_SOURCE_1',
 'EXT_SOURCE_2',
 'EXT_SOURCE_3',
 'APARTMENTS_AVG',
 'BASEMENTAREA_AVG',
 'YEARS_BEGINEXPLUATATION_AVG',
 'YEARS_BUILD_AVG',
 'COMMONAREA_AVG',
 'ELEVATORS_AVG',
 'ENTRANCES_AVG',
 'FLOORSMAX_AVG',
 'FLOORSMIN_AVG',
 'LANDAREA_AVG',
 'LIVINGAPARTMENTS_AVG',
 'LIVINGAREA_AVG',
 'NONLIVINGAPARTMENTS_AVG',
 

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 356251 entries, 0 to 356250
Columns: 770 entries, SK_ID_CURR to CC_COUNT
dtypes: bool(133), float64(584), int64(42), object(11)
memory usage: 1.7+ GB


#Les valeurs manquantes

In [8]:
#le pourcentage de valeurs manquantes pour chaque colonne
valeur_m_p = df.isnull().mean() * 100

#les colonnes avec plus de 70% de valeurs manquantes
valeur_m_cols = valeur_m_p[valeur_m_p > 70].index

#le nombre de colonnes avec plus de 70% de valeurs manquantes
nb_valeur_m_cols = valeur_m_cols.shape[0]
print(f"Le nombre de colonnes avec plus de 70% de valeurs manquantes: {nb_valeur_m_cols}")

Le nombre de colonnes avec plus de 70% de valeurs manquantes: 144


In [ ]:
#le nombre de valeurs manquantes par colonne
missing_values = df.isnull().sum()
missing_values = missing_values[missing_values > 0]
print(missing_values)

TARGET                                  48744
AMT_ANNUITY                                36
AMT_GOODS_PRICE                           278
DAYS_EMPLOYED                           64648
OWN_CAR_AGE                            235239
                                        ...  
CC_NAME_CONTRACT_STATUS_Signed_MAX     252693
CC_NAME_CONTRACT_STATUS_Signed_MEAN    252693
CC_NAME_CONTRACT_STATUS_Signed_SUM     252693
CC_NAME_CONTRACT_STATUS_Signed_VAR     253385
CC_COUNT                               252693
Length: 590, dtype: int64


In [9]:
#la suppression des colonnes avec plus de 70% de valeurs manquantes
df = df.drop(columns=valeur_m_cols)

print(f"Le nombre de colonnes restantes après suppression: {df.shape[1]}")
print(f"Le nombre de colonnes supprimées: {len(valeur_m_cols)}")

Le nombre de colonnes restantes après suppression: 626
Le nombre de colonnes supprimées: 144


In [10]:
#je ne garde que les donneées où la colonne 'TARGET' n'est pas vide.
# 'TARGET' est la variable à prédire (classification), donc ces données sont nécessaires pour l'entraînement.
df_classification = df[df['TARGET'].notnull()]
print(df_classification.shape)

(307507, 626)


In [11]:
#les colonnes cat
cat_cols = df_classification.select_dtypes(include=['object']).columns
print(cat_cols)

Index([], dtype='object')


Après la suppression des cols à >70% de valeurs manquantes, il ne reste plus de cols cat.

#Imputation

In [12]:
#je remplace les valeurs manquantes par la moyenne des colonnes numériques
df_classification_imputed = df_classification.copy()
df_classification_imputed.fillna(df_classification.mean(), inplace=True)

In [13]:
#vérification qu'il n'y a plus de valeurs manquantes
print(df_classification_imputed.isnull().sum().sum())

0


In [14]:
#le df avec imputation mean en csv
df_classification_imputed.to_csv('df_classification_imputed.csv', index=False)

In [ ]:
df = pd.read_csv('df_classification_imputed.csv')
df.shape

#Echantillon


Dans mon df la target est déséquilibrée : environ 90% sont de la classe 0 (négatif), et seulement 10% sont de la classe 1 (positif).

Donc mon modèle pourrait prédire la majorité comme étant de la classe 0 sans optimiser la performance, surtout pour la classe 1


Pour mieux gérer ce déséquilibre, je vais utiliser deux approches :


**SMOTE (Synthetic Minority Over-sampling Technique)** : c'est une technique d'augmentation de données qui permet de créer des données synthétiques de la classe minoritaire (ici 1) pour équilibrer les classes.

**F-beta score (avec β=10)** : c'est une métrique d'évaluation qui permet de pénalizer davantage les faux négatifs (FN, erreurs coûteuses).

___________________________________________________

C'est pour mieux gérer le déséquilibre de classes et optimiser les performances des modèles sur la classe minoritaire


In [23]:
import pandas as pd

#la taille de l'échantillon total
sample_size = 50000

#je calcule la proportion des classes
class_distribution = df_classification_imputed['TARGET'].value_counts(normalize=True)

#échantillon stratifié en fonction de la distribution de 'TARGET'
df_sample = df_classification_imputed.groupby('TARGET', group_keys=False).apply(lambda x: x.sample(int(sample_size * class_distribution.loc[x.name]), random_state=42))

#la distribution dans l'échantillon
print(df_sample['TARGET'].value_counts(normalize=True))


TARGET
0.0    0.919278
1.0    0.080722
Name: proportion, dtype: float64


- TARGET = 1 -> Client avec des difficultés de payments
- TARGET = 0 -> Client sans difficultés de payments

In [2]:
df_sample.to_csv('/content/df_sample.csv', index=False)

In [43]:
import pandas as pd

#pour obtenir un échantillon stratifié en fonction de la distribution de la classe

# La taille de l'échantillon total
sample_size = 50000

# Calcul de la proportion des classes
class_distribution = df_classification_imputed['TARGET'].value_counts(normalize=True)

# Échantillon stratifié en fonction de la distribution de 'TARGET'
df_sample = df_classification_imputed.groupby('TARGET', group_keys=False).apply(lambda x: x.sample(int(sample_size * class_distribution.loc[x.name]), random_state=42))

# Distribution dans l'échantillon
print(df_sample['TARGET'].value_counts(normalize=True))


TARGET
0.0    0.919278
1.0    0.080722
Name: proportion, dtype: float64


#Entrainement des modeles avec tracking MLFlow

In [3]:
import pandas as pd

df_sample = pd.read_csv('/content/df_sample.csv')
df_sample.shape

(49999, 626)

In [4]:
#lightGBM ne prend pas en charge les caractères spéciaux dans les noms de colonnes donc je les supprime
import re

#la fonction pour nettoyer les noms de colonnes
def clean_column_names(df):
    df.columns = [re.sub(r'\W+', '_', col) for col in df.columns]
    return df

#j'applique la fonction à X_train et X_val
df_sample = clean_column_names(df_sample)


#maintenant les colonnes ne contiennent plus de caractères spéciaux

##SMOTE

J'utilise **X_train_resampled et y_train_resampled** pour l'**entraînement**.

J'utilise **X_val et y_val (sans SMOTE)** pour la **validation**, car il est important de ne pas modifier l'ensemble de validation avec SMOTE.

In [5]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE

#Une fois que j'ai df_sample, j'applique SMOTE pour équilibrer les classes

# Séparer les features et la cible
X = df_sample.drop(columns=["TARGET", "SK_ID_CURR"])
y = df_sample["TARGET"]

# Diviser les données en ensembles d'entraînement et de validation
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Application de SMOTE sur l'ensemble d'entraînement
smote = SMOTE(sampling_strategy='auto', random_state=42)
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

# Afficher la nouvelle distribution des classes après SMOTE
print(pd.Series(y_train_resampled).value_counts(normalize=True))

# Vérifier les dimensions des ensembles
print("X_train_resampled shape:", X_train_resampled.shape)
print("y_train_resampled shape:", y_train_resampled.shape)
print("X_val shape:", X_val.shape)
print("y_val shape:", y_val.shape)


TARGET
0.0    0.5
1.0    0.5
Name: proportion, dtype: float64
X_train_resampled shape: (73540, 624)
y_train_resampled shape: (73540,)
X_val shape: (10000, 624)
y_val shape: (10000,)


In [6]:
#j'install MLFlow
! pip install mlflow

##Random Forest Classifier

In [29]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, roc_auc_score
import mlflow
import mlflow.sklearn

###random_forest_default

In [32]:
#l'expérience MLFlow : RandomForestClassifier
mlflow.set_experiment("RandomForestClassifier")

# RUN 1 : random_forest_default

#initialisation du RandomForestClassifier avec hyperparamètres par défaut
rf_clf = RandomForestClassifier()

with mlflow.start_run(run_name='random_forest_default'):
    #l'enregistrement des hyperparamètres
    mlflow.log_param("model_type", "RandomForestClassifier")
    mlflow.log_param("n_estimators", 100)
    mlflow.log_param("max_depth", None)
    mlflow.log_param("random_state", None)

    #l'entraînement du modèle
    rf_clf.fit(X_train_resampled, y_train_resampled)

    #la prédictions
    y_pred_rf = rf_clf.predict(X_val)
    y_prob_rf = rf_clf.predict_proba(X_val)[:, 1]

    #les métriques
    try:
        auc_score_rf = roc_auc_score(y_val y_prob_rf)
    except ValueError:
        auc_score_rf = float('nan')

    accuracy_rf = accuracy_score(y_val, y_pred_rf)

    #l'enregistrement des métriques
    mlflow.log_metric("auc", auc_score_rf)
    mlflow.log_metric("accuracy", accuracy_rf)

    # Enregistrer le modèle dans MLflow (avec signature et input_example pour faciliter le serving)
    model_uri = mlflow.sklearn.log_model(
        rf_clf,
        artifact_path="random_forest_model",
        #input_example=X_val[:5]  # Exemple d'entrée pour la prédiction
    )

    # Enregistrer le modèle dans le Model Registry
    model_name = "RandomForestModel"  # Nom du modèle dans le registre
    mlflow.register_model(model_uri=model_uri.model_uri, name=model_name)

    # Promouvoir le modèle vers 'Staging'
    client = mlflow.tracking.MlflowClient()
    client.transition_model_version_stage(
        name=model_name,
        version=1,  # Vous pouvez choisir la version manuellement ou automatiser
        stage="Staging"  # Les stages peuvent être 'None', 'Staging', 'Production' ou 'Archived'
    )



#pour vérification
print(f"Random Forest AUC: {auc_score_rf}, Accuracy: {accuracy_rf}")


2024/09/27 07:52:45 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
Successfully registered model 'RandomForestModel'.
Created version '1' of model 'RandomForestModel'.
<ipython-input-32-e9d7aa98dc85>:48: FutureWarning: ``mlflow.tracking.client.MlflowClient.transition_model_version_stage`` is deprecated since 2.9.0. Model registry stages will be removed in a future major release. To learn more about the deprecation of model registry stages, see our migration guide here: https://mlflow.org/docs/latest/model-registry.html#migrating-from-stages
  client.transition_model_version_stage(


Random Forest AUC: 0.6948609678367693, Accuracy: 0.9193


In [33]:
#pour vérifier que mon modèle est bien enregistré sur le dossier MLflow (Drive)
!ls

df_classification_imputed.csv  mlruns  projet7_1.ipynb	README.md


###random_forest_hyper1

In [37]:
#l'expérience MLFlow : RandomForestClassifier
mlflow.set_experiment("RandomForestClassifier")

# RUN 2 : random_forest_hyper1

#initialisation du RandomForestClassifier avec hyperparamètres
rf_clf_h1 = RandomForestClassifier(n_estimators=200, max_depth=5, random_state=42)

with mlflow.start_run(run_name='random_forest_hyper1'):
    #l'enregistrement des hyperparamètres
    mlflow.log_param("model_type", "RandomForestClassifier")
    mlflow.log_param("n_estimators", 200)
    mlflow.log_param("max_depth", 5)
    mlflow.log_param("random_state", 42)

    #l'entraînement du modèle
    rf_clf_h1.fit(X_train, y_train)

    #la prédictions
    y_pred_rf_h1 = rf_clf_h1.predict(X_val)
    y_prob_rf_h1 = rf_clf_h1.predict_proba(X_val)[:, 1]

    #les métriques
    try:
        auc_score_rf_h1 = roc_auc_score(y_val, y_prob_rf_h1)
    except ValueError:
        auc_score_rf_h1 = float('nan')

    accuracy_rf_h1 = accuracy_score(y_val, y_pred_rf_h1)

    #l'enregistrement des métriques
    mlflow.log_metric("auc_h1", auc_score_rf_h1)
    mlflow.log_metric("accuracy_h1", accuracy_rf_h1)

    #l'enregistrement du modèle Random Forest
    mlflow.sklearn.log_model(rf_clf_h1, "random_forest_model_h1")

#pour vérification
print(f"Random Forest AUC_h1: {auc_score_rf_h1}, Accuracy_h1: {accuracy_rf_h1}")

2024/09/27 07:57:47 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


Random Forest AUC_h1: 0.7231446371498382, Accuracy_h1: 0.9193


In [ ]:
#pour vérifier que mon modèle est bien enregistré dans mlruns sur le dossier MLflow (Drive)
!ls

In [ ]:
mlflow.get_experiment_by_name("RandomForestClassifier")


Je vais utilier GridSearchCV : le modèle est entraîné avec toutes les combinaisons d'hyperparamètres spécifiées, et la meilleure combinaison est sélectionnée en fonction de la métrique de score :accuracy pour l'instant.

Ensuite, après avoir trouvé les meilleurs hyperparamètres, les prédictions sont effectuées, les métriques sont calculées

###random_forest_grid_search_cv

###random_forest_grid_search_score

In [ ]:
from sklearn.model_selection import GridSearchCV

#l'expérience MLFlow : RandomForestClassifier
mlflow.set_experiment("RandomForestClassifier")


#les hyperparamètres à explorer
param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'random_state': [42]
}

# RUN 3 : random_forest_grid_search_cv

#initialisation du RandomForestClassifier
rf_clf_cv = RandomForestClassifier()

#la configuration de GridSearchCV
grid_search_rf = GridSearchCV(estimator=rf_clf_cv, param_grid=param_grid_rf, cv=5, scoring='accuracy')

#exécution de GridSearchCV
with mlflow.start_run(run_name='random_forest_grid_search_cv'):
    grid_search_rf.fit(X_train, y_train)

    #meilleurs paramètres et score
    best_params_rf = grid_search_rf.best_params_
    best_score_rf = grid_search_rf.best_score_

    #enregistrement des meilleurs paramètres
    mlflow.log_params(best_params_rf)
    mlflow.log_param("best_score", best_score_rf)

    #la prédictions avec les meilleurs paramètres
    y_pred_rf_cv = grid_search_rf.best_estimator_.predict(X_val)
    y_prob_rf_cv = grid_search_rf.best_estimator_.predict_proba(X_val)[:, 1]

    #calcul des métriques
    auc_score_rf_cv = roc_auc_score(y_val, y_prob_rf_cv)
    accuracy_rf_cv = accuracy_score(y_val, y_pred_rf_cv)

    #enregistrement des métriques
    mlflow.log_metric("auc_cv", auc_score_rf_cv)
    mlflow.log_metric("accuracy_cv", accuracy_rf_cv)

    #enregistrement du modèle RandomForest avec les meilleurs paramètres
    mlflow.sklearn.log_model(grid_search_rf.best_estimator_, "random_forest_best_model")

print(f"Best Random Forest Parameters: {best_params_rf}")
print(f"Random Forest AUC_cv: {auc_score_rf_cv}, Accuracy_cv: {accuracy_rf_cv}")

penalized_score = fp + 10 * fn

Je calcule le score en pénalisant les faux négatifs plus fortement

In [ ]:
from sklearn.metrics import confusion_matrix

#l'expérience MLFlow : RandomForestClassifier
mlflow.set_experiment("RandomForestClassifier")

#les hyperparamètres à explorer
param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'random_state': [42]
}

# RUN 4
#initialisation du RandomForestClassifier
rf_clf = RandomForestClassifier()

#la configuration de GridSearchCV
grid_search_rf = GridSearchCV(estimator=rf_clf, param_grid=param_grid_rf, cv=5, scoring='accuracy')

#exécution de GridSearchCV
with mlflow.start_run(run_name='random_forest_grid_search_score'):
    grid_search_rf.fit(X_train, y_train)

    # Meilleurs paramètres et score
    best_params_rf = grid_search_rf.best_params_
    best_score_rf = grid_search_rf.best_score_

    # Enregistrement des meilleurs paramètres
    mlflow.log_params(best_params_rf)
    mlflow.log_param("best_score", best_score_rf)

    # Prédictions avec les meilleurs paramètres
    y_pred_rf = grid_search_rf.best_estimator_.predict(X_val)

    # Calcul des métriques
    confusion = confusion_matrix(y_val, y_pred_rf)
    fp = confusion[0, 1]  # Faux positifs
    fn = confusion[1, 0]  # Faux négatifs

    penalized_score = fp + 10 * fn

    try:
        auc_score_rf = roc_auc_score(y_val, y_pred_rf)
    except ValueError:
        auc_score_rf = float('nan')

    accuracy_rf = accuracy_score(y_val, y_pred_rf)

    # Enregistrement des métriques
    mlflow.log_metric("auc", auc_score_rf)
    mlflow.log_metric("accuracy", accuracy_rf)
    mlflow.log_metric("penalized_score", penalized_score)

    # Enregistrement du modèle RandomForest avec les meilleurs paramètres
    mlflow.sklearn.log_model(grid_search_rf.best_estimator_, "random_forest_best_model_score")

print(f"Best Random Forest Parameters: {best_params_rf}")
print(f"Random Forest AUC: {auc_score_rf}, Accuracy: {accuracy_rf}, Penalized Score: {penalized_score}")

##Light GBM

###lightgbm_default

In [9]:
import lightgbm as lgb
from sklearn.metrics import accuracy_score, roc_auc_score, fbeta_score
import mlflow
import mlflow.sklearn
from mlflow.models import ModelSignature
from mlflow.types.schema import Schema, ColSpec
import os

# Initialisation du LightGBM avec hyperparamètres par défaut
lgb_clf = lgb.LGBMClassifier()

# MLflow expérience
mlflow.set_experiment("LightGBM_default_SMOTE")

with mlflow.start_run(run_name='lightgbm_with_SMOTE_and_fbeta'):
    # Enregistrement des hyperparamètres
    mlflow.log_param("model_type", "LGBMClassifier")
    mlflow.log_param("boosting_type", lgb_clf.boosting_type)
    mlflow.log_param("num_leaves", lgb_clf.num_leaves)
    mlflow.log_param("learning_rate", lgb_clf.learning_rate)
    mlflow.log_param("n_estimators", lgb_clf.n_estimators)
    mlflow.log_param("max_depth", lgb_clf.max_depth)

    # Entraînement du modèle sur les données resamplées
    lgb_clf.fit(X_train_resampled, y_train_resampled)

    # Prédictions sur le jeu de validation (non-resamplé)
    y_pred_lgb = lgb_clf.predict(X_val)
    y_prob_lgb = lgb_clf.predict_proba(X_val)[:, 1]

    # Calcul des métriques
    try:
        auc_score_lgb = roc_auc_score(y_val, y_prob_lgb)
    except ValueError:
        auc_score_lgb = float('nan')

    accuracy_lgb = accuracy_score(y_val, y_pred_lgb)
    fbeta_lgb = fbeta_score(y_val, y_pred_lgb, beta=10)

    # Enregistrement des métriques
    mlflow.log_metric("auc", auc_score_lgb)
    mlflow.log_metric("accuracy", accuracy_lgb)
    mlflow.log_metric("fbeta_score", fbeta_lgb)

    # Enregistrement du modèle dans MLflow (modèle temporaire)
    model_uri = "lightgbm_model_with_SMOTE_and_fbeta"
    mlflow.sklearn.log_model(lgb_clf, model_uri)

    # Enregistrement du modèle dans le Model Registry
    model_name = "LightGBM_with_SMOTE_Model"
    registered_model_uri = mlflow.register_model(model_uri=f"runs:/{mlflow.active_run().info.run_id}/{model_uri}",
                                                 name=model_name)

# Résultats
print(f"LightGBM AUC: {auc_score_lgb}, Accuracy: {accuracy_lgb}, F-beta Score (β=10): {fbeta_lgb}")


2024/09/27 14:07:39 INFO mlflow.tracking.fluent: Experiment with name 'LightGBM_default_SMOTE' does not exist. Creating a new experiment.


[LightGBM] [Info] Number of positive: 36770, number of negative: 36770
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.879243 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 115006
[LightGBM] [Info] Number of data points in the train set: 73540, number of used features: 608
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000


2024/09/27 14:08:30 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


LightGBM AUC: 0.751969435286344, Accuracy: 0.9179, F-beta Score (β=10): 0.042509469931420366


Successfully registered model 'LightGBM_with_SMOTE_Model'.
Created version '1' of model 'LightGBM_with_SMOTE_Model'.


**Conclusion :**

**Équilibrage des classes :**  

Number of positive: 36770

Number of negative: 36770

Cela indique que SMOTE a bien fonctionné


**Performance du modèle :**

**AUC de 0.75 :**

Ce score AUC (Area Under the Curve) indique une performance de classification correcte, mais pas exceptionnelle.

Un score de 0.75 signifie que le modèle est capable de séparer correctement les classes positives et négatives dans environ 75 % des cas.


**Accuracy de 91.79% :**

Le modèle est très précis, mais cela peut être dû au déséquilibre des classes dans l'ensemble de validation, même après SMOTE sur l'entraînement.

Il y a probablement beaucoup plus de classes négatives que positives, ce qui conduit à une haute précision.

**F-beta Score (β=10) de 0.042 :**

Ce score est très faible, ce qui montre que le modèle a des difficultés à identifier les classes positives.

Le F-beta avec β=10 pénalise fortement les faux négatifs, et un faible score ici suggère que le modèle manque beaucoup de véritables positifs.


**Version du modèle :** Une nouvelle version du modèle a été créée dans le Model Registry sous le nom LightGBM_with_SMOTE_Model, et c'est maintenant la version 3.

**Conclusion générale :**

Bien que le modèle affiche une haute précision et un AUC décent, son très faible F-beta score indique qu'il n'est pas bien adapté pour gérer le déséquilibre des classes lors de la prédiction des positifs.

###lightgbm_hyper

In [ ]:
#l'expérience MLFlow : LightGBM
mlflow.set_experiment("LightGBM")

# RUN 2: lightgbm_hyper

#initialisation du LightGBM avec des hyperparamètres personnalisés
lgb_clf_h1 = lgb.LGBMClassifier(
    num_leaves=50,          #j'augmenter le nombre de feuilles
    max_depth=10,           #la profondeur des arbres
    learning_rate=0.05,     #je diminue le taux d'apprentissage
    n_estimators=200        #j'augmenter le nombre d'estimations (arbres)
)

with mlflow.start_run(run_name='lightgbm_custom_h1'):
    #enregistrement des hyperparamètres personnalisés
    mlflow.log_param("model_type", "LGBMClassifier")
    mlflow.log_param("boosting_type", lgb_clf_h1.boosting_type)  #c'était 'gbdt' par défaut
    mlflow.log_param("num_leaves", lgb_clf_h1.num_leaves)
    mlflow.log_param("learning_rate", lgb_clf_h1.learning_rate)
    mlflow.log_param("n_estimators", lgb_clf_h1.n_estimators)
    mlflow.log_param("max_depth", lgb_clf_h1.max_depth)

    #entraînement du modèle
    lgb_clf_h1.fit(X_train, y_train)

    #la prédictions
    y_pred_lgb_h1 = lgb_clf_h1.predict(X_val)
    y_prob_lgb_h1 = lgb_clf_h1.predict_proba(X_val)[:, 1]

    #je calcul les métriques
    try:
        auc_score_lgb_h1 = roc_auc_score(y_val, y_prob_lgb_h1)
    except ValueError:
        auc_score_lgb_h1 = float('nan')

    accuracy_lgb_h1 = accuracy_score(y_val, y_pred_lgb_h1)

    #l'enregistrement des métriques
    mlflow.log_metric("auc_h1", auc_score_lgb_h1)
    mlflow.log_metric("accuracy_h1", accuracy_lgb_h1)

    #l'enregistrement du modèle LightGBM
    mlflow.sklearn.log_model(lgb_clf_h1, "lightgbm_custom_model_h1")

#les résultats pour vérification
print(f"LightGBM AUC_h1: {auc_score_lgb_h1}, Accuracy_h1: {accuracy_lgb_h1}")


###lightgbm_grid_search_cv

In [ ]:
#l'expérience MLFlow : LightGBM
mlflow.set_experiment("LightGBM")

#les hyperparamètres à explorer
param_grid_lgb = {
    'num_leaves': [31, 50, 70],
    'max_depth': [-1, 10, 20],
    'learning_rate': [0.01, 0.05, 0.1],
    'n_estimators': [100, 200, 300]
}

#initialisation du LGBMClassifier
lgb_clf_cv = lgb.LGBMClassifier()

# RUN 3

#configuration de GridSearchCV
grid_search_lgb = GridSearchCV(estimator=lgb_clf_cv, param_grid=param_grid_lgb, cv=5, scoring='accuracy')

#exécution de GridSearchCV
with mlflow.start_run(run_name='lightgbm_grid_search_cv'):
    grid_search_lgb.fit(X_train, y_train)

    #meilleurs paramètres et score
    best_params_lgb = grid_search_lgb.best_params_
    best_score_lgb = grid_search_lgb.best_score_

    #enregistrement des meilleurs paramètres
    mlflow.log_params(best_params_lgb)
    mlflow.log_param("best_score", best_score_lgb)

    #prédictions avec les meilleurs paramètres
    y_pred_lgb_cv = grid_search_lgb.best_estimator_.predict(X_val)
    y_prob_lgb_cv = grid_search_lgb.best_estimator_.predict_proba(X_val)[:, 1]

    #calcul des métriques
    auc_score_lgb_cv = roc_auc_score(y_val, y_prob_lgb_cv)
    accuracy_lgb_cv = accuracy_score(y_val, y_pred_lgb_cv)

    #enregistrement des métriques
    mlflow.log_metric("auc_cv", auc_score_lgb_cv)
    mlflow.log_metric("accuracy_cv", accuracy_lgb_cv)

    #enregistrement du modèle LGBM avec les meilleurs paramètres
    mlflow.sklearn.log_model(grid_search_lgb.best_estimator_, "lightgbm_best_model")

print(f"Best LightGBM Parameters: {best_params_lgb}")
print(f"LightGBM AUC_cv: {auc_score_lgb_cv}, Accuracy_cv: {accuracy_lgb_cv}")


###lightgbm_GridSearchCV

In [8]:
import lightgbm as lgb
from sklearn.metrics import accuracy_score, roc_auc_score, fbeta_score
from sklearn.model_selection import GridSearchCV
import mlflow
import mlflow.sklearn
from mlflow.models import ModelSignature
from mlflow.types.schema import Schema, ColSpec
import os

# Définir la grille d'hyperparamètres à tester
param_grid = {
    'num_leaves': [31, 50, 70],
    'learning_rate': [0.01, 0.05, 0.1],
    'n_estimators': [100, 200, 500],
    'max_depth': [-1, 10, 20],
    'min_child_samples': [20, 50, 100]
}

# Initialisation du LightGBM avec hyperparamètres par défaut
lgb_clf = lgb.LGBMClassifier()

# GridSearchCV pour optimiser les hyperparamètres
grid_search = GridSearchCV(estimator=lgb_clf, param_grid=param_grid, scoring='roc_auc', cv=5, n_jobs=-1, verbose=1)

# MLflow expérience
mlflow.set_experiment("LightGBM_GridSearchCV")

with mlflow.start_run(run_name='lightgbm_with_GridSearchCV'):
    # Enregistrement des hyperparamètres avant GridSearch
    mlflow.log_param("model_type", "LGBMClassifier")

    # Entraînement avec GridSearchCV sur les données resamplées
    grid_search.fit(X_train_resampled, y_train_resampled)

    # Obtenir les meilleurs hyperparamètres
    best_params = grid_search.best_params_
    print(f"Meilleurs hyperparamètres trouvés : {best_params}")

    # Meilleur modèle entraîné : modèle entraîné avec les meilleurs hyperparamètres (best_lgb_model)
    best_lgb_model = grid_search.best_estimator_

    # Enregistrement des hyperparamètres optimisés
    mlflow.log_params(best_params)

    # Prédictions sur le jeu de validation (non-resamplé)
    y_pred_lgb = best_lgb_model.predict(X_val)
    y_prob_lgb = best_lgb_model.predict_proba(X_val)[:, 1]

    # Calcul des métriques
    try:
        auc_score_lgb = roc_auc_score(y_val, y_prob_lgb)
    except ValueError:
        auc_score_lgb = float('nan')

    accuracy_lgb = accuracy_score(y_val, y_pred_lgb)
    fbeta_lgb = fbeta_score(y_val, y_pred_lgb, beta=10)

    # Enregistrement des métriques
    mlflow.log_metric("auc", auc_score_lgb)
    mlflow.log_metric("accuracy", accuracy_lgb)
    mlflow.log_metric("fbeta_score", fbeta_lgb)

    # Enregistrement du meilleur modèle dans MLflow (modèle temporaire)
    model_uri = "lightgbm_model_with_SMOTE_and_fbeta"
    mlflow.sklearn.log_model(best_lgb_model, model_uri)

    # Enregistrement du modèle dans le Model Registry
    model_name = "LightGBM_with_SMOTE_Model"
    registered_model_uri = mlflow.register_model(model_uri=f"runs:/{mlflow.active_run().info.run_id}/{model_uri}",
                                                 name=model_name)

# Résultats finaux
print(f"LightGBM AUC: {auc_score_lgb}, Accuracy: {accuracy_lgb}, F-beta Score (β=10): {fbeta_lgb}")

#MLFlow UI

In [34]:
#maintenant pour visualiser sur mlflw j'ai besoin de ngrok
#j'installe le package
! pip install pyngrok

In [61]:
from pyngrok import ngrok
#terminate open tunnels if exist
ngrok.kill()

#setting the authtoken (optional)
NGROK_AUTH_TOKEN = "2m0LsyXXeuXcuZ6p34hmutJGy1u_4yK2gMFBwyn1ka2HPfR1i"
ngrok.set_auth_token(NGROK_AUTH_TOKEN)

#open an HTTPs tunnel on port 5000 for http://localhost:5000
ngrok_tunnel = ngrok.connect(addr = "5000", proto= "http", bind_tls =True)
print('MLFlow Tracking UI:', ngrok_tunnel.public_url)

MLFlow Tracking UI: https://4955-34-32-209-47.ngrok-free.app


In [62]:
#afin de visualiser les expérimentations avec MLFlow, j'exécute l'interface utilisateur
!mlflow ui

[2024-09-27 12:55:44 +0000] [83983] [INFO] Starting gunicorn 23.0.0
[2024-09-27 12:55:44 +0000] [83983] [INFO] Listening at: http://127.0.0.1:5000 (83983)
[2024-09-27 12:55:44 +0000] [83983] [INFO] Using worker: sync
[2024-09-27 12:55:44 +0000] [83988] [INFO] Booting worker with pid: 83988
[2024-09-27 12:55:44 +0000] [83989] [INFO] Booting worker with pid: 83989
[2024-09-27 12:55:44 +0000] [83990] [INFO] Booting worker with pid: 83990
[2024-09-27 12:55:44 +0000] [83991] [INFO] Booting worker with pid: 83991
[2024-09-27 12:59:47 +0000] [83983] [INFO] Handling signal: int

Aborted!
[2024-09-27 12:59:47 +0000] [83991] [INFO] Worker exiting (pid: 83991)
[2024-09-27 12:59:47 +0000] [83989] [INFO] Worker exiting (pid: 83989)
[2024-09-27 12:59:47 +0000] [83988] [INFO] Worker exiting (pid: 83988)
[2024-09-27 12:59:47 +0000] [83990] [INFO] Worker exiting (pid: 83990)
[2024-09-27 12:59:49 +0000] [83983] [INFO] Shutting down: Master
